openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+        structure.
+        """
+        return self._block_relations.iterkeys()
+
+    def _add_relation(self, parent_key, child_key):
+        """
+        Adds a parent to child relationship in this block structure.
+
+        Arguments:
+            parent_key (UsageKey) - Usage key of the parent block.
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        self._add_to_relations(self._block_relations, parent_key, child_key)
+
+    @staticmethod
+    def _add_to_relations(block_relations, parent_key, child_key):
+        """
+        Adds a parent to child relationship in the given block
+        relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            parent_key (UsageKey) - Usage key of the parent block.
+
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        block_relations[child_key].parents.append(parent_key)
+        block_relations[parent_key].children.append(child_key)
+
+    @staticmethod
+    def _add_block(block_relations, usage_key):
+        """
+        Adds the given usage_key to the given block_relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            usage_key (UsageKey) - Usage key of the block that is to
+                be added to the given block_relations.
+        """
+
+        # Since block_relations is a DefaultDict, the usage_key will
+        # automatically be added by virtue of accessing it.
+        _ = block_relations[usage_key]
+
+
+class _BlockData(object):
+    """
+    Data structure to encapsulate collected data for a single block.
+    """
+    def __init__(self):
+        # Map of xblock field name to the field's value for this block.
+        # dict {string: any picklable type}
+        self._xblock_fields = {}
+
+        # Map of transformer name to the transformer's data for this
+        # block.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
+
+
+class BlockStructureBlockData(BlockStructure):
+    """
+    Subclass of BlockStructure that is responsible for managing block
+    and transformer data.
+    """
+    def __init__(self, root_block_usage_key):
+        super(BlockStructureBlockData, self).__init__(root_block_usage_key)
+
+        # Map of a block's usage key to its collected data, including
+        # its xBlock fields and block-specific transformer data.
+        # defaultdict {UsageKey: _BlockData}
+        self._block_data_map = defaultdict(_BlockData)
+
+        # Map of a transformer's name to its non-block-specific data.
+        # defaultdict {string: dict}
+        self._transformer_data = defaultdict(dict)
```

Comment:
On the other hand, those attributes are private in a public class.  No need to document them.